### ISSUE TITLE
[Feature] Tightly couple top level element with the styled component 

### ISSUE THREAD

**Issue by danielrob** [styled-components/styled-components#1702]
With [React.Fragment](https://reactjs.org/docs/fragments.html) it might be possible to do something like:
```
import React from 'react'
import styled from 'styled-components'

const People = () => (<>
  <Friends />
  <Foes />
</>)


export default styled(People).div`
  display: flex;
`
```
Note the `styled(People).div` which would render e.g. 
```
<div className="sc-a2ad2d a2ad2d" {...whitelistedProps}>
  <People {...props} />
</div>
```

## Presently there are two patterns: 
The recommended pattern
```
const PeopleWrapper = styled.div`
  display: flex;
`

const People = () => (
    <PeopleWrapper>
      <Friends />
      <Foes />       
    </PeopleWrapper>
)

```
And the other pattern: 
```
const People = ({ className }) => (
    <div className={className}>
      <Friends />
      <Foes />       
    </div>
)

export default styled(People)`
  display: flex;
`
```

### Shortfalls:
The recommended approach: 
 -  requires you to name the styling of the outer node of your component (e.g. `PeopleWrapper` ) even if it's only purpose is to attach a styling wrapper to the components outer element. Naming things is really hard and even if it's deciding on Wrapper once and everywhere having two names for the two halves of one component is really awful. The whole point of a component is that it encapsulates everything including style and function.

The non-recommended approach: 
 -  still contains some boilerplate, and does not disallow developers to mistakenly pass non-whitelisted props to the outer-node, among other things. 

### Potential wins: 
In reality, compared to the approach being requested, neither of these shortfalls is _that_ bad, but  using `React.Fragment` could allow us to have a universal pattern for naming our component just once, and yet still bundle our styles with it, regardless of whether the top level node was intended to contain one or more children. 

Basically, I think this syntax (if possible) is really freaking nice. I
 - don't have to name a wrapper. 
 - maintain the reversed order of consumption which allows me to specify my styles underneath my declared component in the same file, without requiring hoisting. 
 - have more explicitly the understanding that when using styled components, the component has a styled outer element. Presently with the added step of naming my wrapper and calling it this desired coupling feels less coupled than it could. 
 - focus on the content being placed in the component without the distraction of the 'style attachment' declaration that is currently the first item in all components of this type. 

These add up to a very lean and clean refinement of an already beautiful api in my books :). 

### Implementation
Not the expert here, but I'd hope the fact that presently `styled(Component)` spits out a function that expects `(['text', 'text'], ...args)` it would be simple enough to pass the `.element` helpers on as properties to the result of that factory function, and not just attach them to the factory function themselves? 

Note also that React.Fragment has nothing to do with how this would be implemented, but simply that React.Fragment could replace the outer node when the outer node was previously serving a dual purpose as a wrapper for multiple children as well as being an attachment element for styling.  It would not be required by the end developer when their component renders e.g. `<People><Friends /></People>`

**Comment by kitten** [comment]
`styled(Component)` is not actually the recommended pattern at all. If you have your own components that you're styling, you might want to instead style elements, and not containers. Containers, be it stateful or -less can simply provide props for switching styles. Often it's also not necessary to even provide a lot of props, or avoid just creating more elements with styles.

But even then, I don't think this would accomplish much. It would undermine what fragments are supposed to do. `styled(Component).div` is not a thing and that's for a reason. Here you're basically saying: Use this component, with this element.

That semantically doesn't make sense. It's not intuitively answering: Where does the className go? Where does the element go? Is it just being wrapped?

So basically, while this is an API that you could code as a utility and separate lib—which i'd recommend—from a core library's stand point it's an API that would recommend the usage of a pattern that we'd like to be a "second class" pattern

**Comment by danielrob** [comment]
@kitten thanks for this! In fact I think I've just finished up the source code for this secondary lib! It's late for me now, but I'll post it here / on github / on npm tomorrow :) 

I had previously read the long threads on the 'as any element' argument, and I actually strongly supported your decision to disallow that. Ironically though when I've come at it from this different angle, I like the possibility precisely because 'stating where className goes' ends up being boilerplate for me in many cases. It seems 'styled-components' is more or less a 'styled-elements' library in truth; with an escape hatch.

**Comment by danielrob** [comment]
@kitten and voila:
https://github.com/danielrob/styled-as-components/ 
https://www.npmjs.com/package/styled-as-components

Please feel free to tell me I've done something very evil. This is my first public npm package :).
